src/aiida_wannier90_workflows/utils/pseudo/__init__.py

```diff
@@ -259,20 +405,68 @@ def get_number_of_projections(
     for site in structure.sites:
         upf = pseudos[site.kind_name]
         nprojs = get_number_of_projections_from_upf(upf)
-        soc = is_soc_pseudo(get_upf_content(pseudos[site.kind_name]))
-        if spin_orbit_coupling and not soc:
-            # For SOC calculation with non-SOC pseudo, QE will generate
+        if spin_non_collinear and not spin_orbit_coupling:
+            # For magnetic calculation with non-SOC pseudo, QE will generate
             # 2 PSWFCs from each one PSWFC in the pseudo
+            # For collinear-magnetic calculation, spin up and down will calc
+            # seperately, so nprojs do not times 2
             nprojs *= 2
-        elif not spin_orbit_coupling and soc:
-            # For non-SOC calculation with SOC pseudo, QE will average
+        elif not spin_non_collinear and spin_orbit_coupling:
+            # For non-magnetic calculation with SOC pseudo, QE will average
             # the 2 PSWFCs into one
             nprojs //= 2
         tot_nprojs += nprojs
 
     return tot_nprojs
 
 
+def get_number_of_projections_ext(
+    structure: orm.StructureData,
+    external_projectors: dict,
+    spin_non_collinear: bool,
+    spin_orbit_coupling: bool = False,
+) -> int:
+    """Get number of projections for the structure with the given projector dict.
+
+    :param structure: crystal structure
+    :type structure: aiida.orm.StructureData
+    :param projectors: a dictionary contains projector list of the structure
+    :type pseudos: dict
+    :param spin_non_collinear: non-collinear or spin-orbit-coupling
+    :type spin_non_collinear: bool
+    :return: number of projections
+    :rtype: int
+    """
+    if not isinstance(structure, orm.StructureData):
+        raise ValueError(
+            f"The type of structure is {type(structure)}, only aiida.orm.StructureData is accepted"
+        )
+    # e.g. composition = {'Ga': 1, 'As': 1}
+    composition = structure.get_composition()
+
+    # I use the first projector to detect SOCs
+    kind = list(composition.keys())[0]
```

Comment:
This still uses the logic before #60, does the dictionary of external projectors have atom symbols as keys (e.g. "Mn") or kind names (could be e.g. "Mn3d")?
```suggestion
     # I use the first projector to detect SOCs
    kind = structure.get_kind_names()[0]
```